pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -302,6 +306,128 @@ func (m *kubeGenericRuntimeManager) getKubeletContainerStatuses(podSandboxID str
+		return
+	}
+	m.recorder.Event(ref, eventtype, reason, message)
+}
+
+// executePreStopHook runs the pre-stop lifecycle hooks if applicable and returns new gracePeriod.
+func (m *kubeGenericRuntimeManager) executePreStopHook(pod *api.Pod, containerID kubecontainer.ContainerID, containerSpec *api.Container, gracePeriod int64) int64 {
+	start := unversioned.Now()
+	if pod != nil && containerSpec != nil && containerSpec.Lifecycle != nil && containerSpec.Lifecycle.PreStop != nil {
+		glog.V(3).Infof("Running preStop hook for container %q", containerID.String())
+		done := make(chan struct{})
+		go func() {
+			defer close(done)
+			defer utilruntime.HandleCrash()
+			if msg, err := m.runner.Run(containerID, pod, containerSpec, containerSpec.Lifecycle.PreStop); err != nil {
+				glog.Errorf("preStop hook for container %q failed: %v", containerSpec.Name, err)
+				m.generateContainerEvent(containerID, api.EventTypeWarning, events.FailedPreStopHook, msg)
+			}
+		}()
+		select {
+		case <-time.After(time.Duration(gracePeriod) * time.Second):
+			glog.V(2).Infof("preStop hook for container %q did not complete in %d seconds", containerID, gracePeriod)
+		case <-done:
+			glog.V(3).Infof("preStop hook for container %q completed", containerID)
+		}
+		gracePeriod -= int64(unversioned.Now().Sub(start.Time).Seconds())
+	}
+
+	return gracePeriod
+}
+
+// killContainer kills a container through the following steps:
+// * Run the pre-stop lifecycle hooks (if applicable).
+// * Stop the container.
+func (m *kubeGenericRuntimeManager) killContainer(pod *api.Pod, containerID kubecontainer.ContainerID, containerSpec *api.Container, reason string, gracePeriodOverride *int64) error {
+	gracePeriod := int64(minimumGracePeriodInSeconds)
+	if pod != nil {
+		switch {
+		case pod.DeletionGracePeriodSeconds != nil:
+			gracePeriod = *pod.DeletionGracePeriodSeconds
+		case pod.Spec.TerminationGracePeriodSeconds != nil:
+			gracePeriod = *pod.Spec.TerminationGracePeriodSeconds
+		}
+	}
+
+	glog.V(2).Infof("Killing container %q with %d second grace period", containerID.String(), gracePeriod)
+
+	// Run the pre-stop lifecycle hooks if applicable.
+	gracePeriod = m.executePreStopHook(pod, containerID, containerSpec, gracePeriod)
+	if gracePeriodOverride == nil {
+		// always give containers a minimal shutdown window to avoid unnecessary SIGKILLs
+		if gracePeriod < minimumGracePeriodInSeconds {
+			gracePeriod = minimumGracePeriodInSeconds
+		}
+	} else {
+		gracePeriod = *gracePeriodOverride
+		glog.V(3).Infof("Killing container %q, but using %d second grace period override", containerID, gracePeriod)
+	}
+
+	err := m.runtimeService.StopContainer(containerID.ID, gracePeriod)
+	if err != nil {
+		glog.Errorf("Container %q termination failed with gracePeriod %d: %v", containerID.String(), gracePeriod, err)
+	} else {
+		glog.V(3).Infof("Container %q exited normally", containerID.String())
+	}
+
+	if _, ok := m.containerRefManager.GetRef(containerID); !ok {
```

Comment:
You don't need the `if` anymore since you call `generateContainerEvent` which does the same thing. You'll only have to handle `ClearRef` yourself, or add another argument in `generateContainerEvent` to handle the cleanup.